File: lib/features/profiles/domain/failover_group.dart
// A failover group as Flutter sees it (SPEC 10.1). The stored copy lives in
// Kotlin beside the profiles; this is what crosses the profile method channel.
import 'package:equatable/equatable.dart';

/// An ordered set of profiles tried one after another until one comes up.
///
/// The members are profile ids rather than whole profiles. The host sends them
/// that way on purpose: the caller already holds the profile list, and carrying
/// each member twice would make two copies of one profile that could disagree
/// after an edit. Resolving an id that no longer names a profile is therefore
/// this side's job — see [resolveMembers].
class FailoverGroup extends Equatable {
  const FailoverGroup({
    required this.id,
    required this.name,
    this.connectTimeoutSec = defaultConnectTimeoutSec,
    this.createdAt = 0,
    this.memberIds = const <String>[],
  });

  final String id;
  final String name;

  /// How long one member may take before the group gives up on it and tries
  /// the next (SPEC 10.1.2).
  final int connectTimeoutSec;

  /// Milliseconds since the epoch, so the list keeps the order groups were
  /// made in. Zero on a group that has not been saved yet; the host fills it.
  final int createdAt;

  /// Profile ids, in the order the group tries them.
  final List<String> memberIds;

  /// SPEC 10.1.2's default budget for one member.
  static const int defaultConnectTimeoutSec = 15;

  /// The range a budget may take. Mirrors `FailoverGroup.kt`: below the floor a
  /// member is abandoned before a slow network has answered at all, and above
  /// the ceiling a group of three is a longer wait than anyone will sit
  /// through. The host clamps too — this copy is here so the editor can say no
  /// before the round trip rather than silently changing what was typed.
  static const int minConnectTimeoutSec = 5;
  static const int maxConnectTimeoutSec = 120;

  /// [seconds] if it is a budget this application will run, the default
  /// otherwise.
  static int normalizeTimeout(int seconds) {
    return seconds >= minConnectTimeoutSec && seconds <= maxConnectTimeoutSec
        ? seconds
        : defaultConnectTimeoutSec;
  }

  /// The name to show. A group is never saved without one, but a group read
  /// from an older store might still arrive empty.
  String get displayName => name.trim().isEmpty ? id : name.trim();

  bool get isEmpty => memberIds.isEmpty;

  FailoverGroup copyWith({
    String? id,
    String? name,
    int? connectTimeoutSec,
    int? createdAt,
    List<String>? memberIds,
  }) {
    return FailoverGroup(
      id: id ?? this.id,
      name: name ?? this.name,
      connectTimeoutSec: connectTimeoutSec ?? this.connectTimeoutSec,
      createdAt: createdAt ?? this.createdAt,
      memberIds: memberIds ?? this.memberIds,
    );
  }

  Map<String, Object?> toJson() => <String, Object?>{
    'id': id,
    'name': name,
    'connectTimeoutSec': connectTimeoutSec,
    'createdAt': createdAt,
    'memberIds': memberIds,
  };

  /// The group in [raw], or null if it is not one.
  ///
  /// A row without an id is dropped rather than repaired: it cannot be saved
  /// back, deleted, or connected, so showing it would only offer actions that
  /// all fail.
  static FailoverGroup? tryFromJson(Object? raw) {
    if (raw is! Map) return null;
    final id = (raw['id'] as String?)?.trim() ?? '';
    if (id.isEmpty) return null;
    final members = <String>[];
    final rawMembers = raw['memberIds'];
    if (rawMembers is List) {
      for (final entry in rawMembers) {
        final memberId = (entry as String?)?.trim() ?? '';
        if (memberId.isNotEmpty && !members.contains(memberId)) {
          members.add(memberId);
        }
      }
    }
    return FailoverGroup(
      id: id,
      name: (raw['name'] as String?)?.trim() ?? '',
      connectTimeoutSec: normalizeTimeout(
        (raw['connectTimeoutSec'] as num?)?.toInt() ?? defaultConnectTimeoutSec,
      ),
      createdAt: (raw['createdAt'] as num?)?.toInt() ?? 0,
      memberIds: members,
    );
  }

  /// The members of this group as rows of [profiles], in the group's order.
  ///
  /// Ids that name nothing are skipped. A profile can be deleted while a group
  /// still mentions it — the host's foreign key removes the membership row, but
  /// a list already in memory has not heard about it yet — and the shorter list
  /// is the honest answer.
  List<T> resolveMembers<T>(
    Iterable<T> profiles,
    String Function(T profile) idOf,
  ) {
    final byId = <String, T>{
      for (final profile in profiles) idOf(profile): profile,
    };
    final out = <T>[];
    for (final memberId in memberIds) {
      final profile = byId[memberId];
      if (profile != null) out.add(profile);
    }
    return out;
  }

  @override
  List<Object?> get props => [
    id,
    name,
    connectTimeoutSec,
    createdAt,
    memberIds,
  ];
}

File: core-trust/src/main/java/io/github/mr1ve3r/combined/core/profile/FailoverGroup.kt
package io.github.mr1ve3r.combined.core.profile

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

/**
 * An ordered set of profiles tried one after another (SPEC 10.1.1).
 *
 * The group holds no connection settings of its own beyond the budget: it is a
 * list of profiles and the order to try them in, and every profile in it is a
 * profile that can also be connected on its own. The case it exists for is one
 * server reachable two ways — L2TP/IPsec first because UDP is quicker, SSTP on
 * 443 behind it for networks where UDP/500 and ESP are blocked.
 *
 * @property connectTimeoutSec how long one member may take before the group
 *   gives up on it and tries the next. SPEC 10.1.2 sets the default at 15,
 *   which is also what makes the acceptance criterion reachable: a fall-through
 *   to the second member has to happen inside 20 seconds.
 * @property createdAt milliseconds since the epoch, so the list the user sees
 *   keeps the order the groups were made in, the way profiles do.
 */
@Entity(tableName = "failover_groups")
data class FailoverGroup(
    @PrimaryKey val id: String,
    val name: String,
    val connectTimeoutSec: Int = DEFAULT_CONNECT_TIMEOUT_SEC,
    val createdAt: Long,
) {
    companion object {
        /** SPEC 10.1.2's default budget for one member. */
        const val DEFAULT_CONNECT_TIMEOUT_SEC: Int = 15

        /**
         * The range a budget may take.
         *
         * The floor is not a style choice: below a couple of seconds a member
         * is abandoned before a slow network has answered at all, which turns a
         * group into a way of failing over past servers that work. The ceiling
         * keeps a group of three inside a wait a user will sit through.
         */
        const val MIN_CONNECT_TIMEOUT_SEC: Int = 5
        const val MAX_CONNECT_TIMEOUT_SEC: Int = 120

        /** [seconds] if it is a budget this application will run, the default otherwise. */
        fun normalizeTimeout(seconds: Int): Int =
            if (seconds in MIN_CONNECT_TIMEOUT_SEC..MAX_CONNECT_TIMEOUT_SEC) seconds else DEFAULT_CONNECT_TIMEOUT_SEC
    }
}

/**
 * One profile's place in one group.
 *
 * The order is a column rather than the insertion order of the rows: a table
 * has no order of its own, and the whole point of a group is that its members
 * are tried in the order the user arranged them.
 *
 * Both foreign keys cascade. Deleting a profile takes it out of every group
 * that named it, which is what a user deleting a profile means; deleting a
 * group takes its membership with it and leaves the profiles alone.
 *
 * @property position 0-based. Gaps are harmless — every read orders by this
 *   column and nothing reads the number itself — but [FailoverGroupStore]
 *   rewrites the whole membership on every change, so they do not arise.
 */
@Entity(
    tableName = "failover_group_member",
    primaryKeys = ["groupId", "profileId"],
    foreignKeys = [
        ForeignKey(
            entity = FailoverGroup::class,
            parentColumns = ["id"],
            childColumns = ["groupId"],
            onDelete = ForeignKey.CASCADE,
        ),
        ForeignKey(
            entity = VpnProfile::class,
            parentColumns = ["id"],
            childColumns = ["profileId"],
            onDelete = ForeignKey.CASCADE,
        ),
    ],
    indices = [Index("profileId")],
)
data class FailoverGroupMember(
    val groupId: String,
    val profileId: String,
    val position: Int,
)

/** A group together with the profiles it names, in the order it names them. */
data class FailoverGroupWithMembers(
    val group: FailoverGroup,
    val members: List<VpnProfile>,
)
